main.tf

```diff
@@ -57,6 +64,43 @@ resource "null_resource" "module_depends_on" {
   }
 }
 
+resource "null_resource" "cache_folder" {
+  count = (var.platform == "" || (var.enabled && ! var.skip_download)) ? 1 : 0
+  provisioner "local-exec" {
+    when    = create
+    command = local.create_cache_folder_cmd
+  }
+  depends_on = [null_resource.module_depends_on]
+}
+
+resource "null_resource" "platform_detect" {
+  count = (var.enabled && ! var.skip_download && var.platform == "") ? 1 : 0
+  triggers = {
+    value = md5(var.platform)
+    value = var.platform == "" && ! fileexists(local.platform_file)
+  }
+  provisioner "local-exec" {
+    when    = create
+    command = local.platform_detect_cmd
+  }
+  depends_on = [
+    null_resource.module_depends_on,
+    null_resource.cache_folder,
+  ]
+}
+
+data "local_file" "platform_detected" {
```

Comment:
Could we add a brief explanation (in a comment) of how this flow is meant to work?

My understanding:
1. On first run, `fileexists` will be *false* so the flow is `null_resource.platform_detect` then call this data source.
2. On subsequent runs, `fileexists` will be *true* so the flow is to directly skip to `copy` and use `file(` to pick up the flow.

This works, but I wonder if it could be simpler. In particular, I'm not sure there's much value in caching the platform detection since `gcloud` scripts should generally only need to be invoked ~once in a particular module. So maybe we could always redetect when we need to run the command?

Reply:
Your understanding is correct.
The reason I ended up doing it like that is because
- I need a bash script to detect the OS
- The only way I found to get back the value is through a file 
- local_file is "dynamically resolved" and can be sequenced with a depends_on, but was  causing the plan to always contain an action (dynamic reading of the file each time).
- fileexists is evaluated at the "planing" time so it's not useable during the first run (the file is not created at plan time), but is useable afterward without generating a plan action.

I'll give it another try to use local_file, without generating an action for each run.
I probably missed something in the triggers